posts/building-a-realtime-app-with-phoenix.md

```diff
@@ -0,0 +1,604 @@
+
+We can handle this on the front-end with very little code, thanks once again to the Presence client-side library. 
+
+```javascript
+import { Socket, Presence } from 'phoenix';
+
+let presences = []
+channel.on("presence_state", (response) => {
+  presences = Presence.syncDiff(presences, response);
+  let participants = Presence.list(presences).map(p => {participants.push(p.metas[0])})
+})
+```
+
+It's pretty much the exact same code we used to handle the broadcast of present users to existing subscribers. The only difference is that this time we're listening for the `"presence_state"` event. 
+
+#### User Leaves a Room
+
+The last event that impacts user presence in a challenge channel is the departure of a user. Once again, Phoenix Presence abstracts away a lot of the work for us. All we have to do to remove a user from the Presence store and broadcast the updated list to the remaining clients is call `channel.leave()` on the front-end. 
+
+This will automatically update the Presence store's list of users and send out a `"presence_diff"` event. Since we're already listening for and handling the `"presence_diff"` event on the front-end, we don't need to write any more code to handle the "user leaves" scenario. 
+
+### Tracking User Typing
+
+Now that we're using Presence to track a user's presence in a given challenge, let's use it to track a custom event: the user typing.
+
+Our `ChallengeChannel` already handles the `"response:update"` event that fires when a user types into the shared text editor. And our Presence store already stores some metadata for a given user that indicates whether or not they are typing. 
+
+Let's hook into the `"response:update"` event in our `ChallengeChannel` to update that portion of a user's metadata in the Presence store. 
+
+```elixir
+# challenge_channel.ex
+
+def handle_in("response:update", %{"response" => response, "user_id" => user_id}, socket) do
+    case Challenge.update(current_challenge(socket), response) do
+      {:ok, challenge} ->
+        ChallengePresence.do_user_update(socket, socket.assigns.current_user, %{typing: true})
+        broadcast! socket, "response:updated", %{challenge: challenge}
+        {:noreply, socket}
+      {:error, changeset} ->
+        {:reply, {:error, %{error: "Error updating challenge"}}, socket}
+    end
+  end
+```
+
+Here's the line we care about:
+
+```elixir
+ ChallengePresence.do_user_update(socket, current_user(socket), %{typing: true})
+```
+
+The `do_user_update` function is a loose wrapper for the `Presence.update` function:
+
+```elixir
+# challenge_presence.ex
+
+def do_user_update(socket, user, %{typing: typing}) do
+  ChallengePresence.update(socket, user.id, %{
+    typing: typing,
```

Comment:
Just curious is there any point, once this is set to true, that it will be reset back to false, may have missed it.